src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/ProjectSystem/RazorProjectInfoEndpointPublisher.cs

```diff
@@ -0,0 +1,194 @@
+    {
+        // Don't do any work if the solution is closing
+        if (args.SolutionIsClosing)
+        {
+            return;
+        }
+
+        switch (args.Kind)
+        {
+            case ProjectChangeKind.ProjectChanged:
+            case ProjectChangeKind.DocumentChanged:
+            case ProjectChangeKind.DocumentRemoved:
+            case ProjectChangeKind.DocumentAdded:
+                if (!ProjectWorkspacePublishable(args.Newer))
+                {
+                    break;
+                }
+
+                EnqueuePublish(args.Newer!);
+
+                break;
+
+            case ProjectChangeKind.ProjectAdded:
+
+                if (ProjectWorkspacePublishable(args.Newer))
+                {
+                    // Don't enqueue project addition as those are unlikely to come through in large batches.
+                    // Also ensures that we won't get project removal go through the queue without project addition.
+                    ImmediatePublish(args.Newer!);
+                }
+
+                break;
+
+            case ProjectChangeKind.ProjectRemoved:
+                // Enqueue removal so it will replace any other project changes in the work queue as they unnecessary now.
+                EnqueueRemoval(args.Older!);
+                break;
+
+            default:
+                Debug.Fail("A new ProjectChangeKind has been added that the RazorProjectInfoEndpointPublisher doesn't know how to deal with");
+                break;
+        }
+    }
+
+    // We want to wait until our project has been initialized (ProjectWorkspaceState != null)
+    // so that we don't publish half-finished projects, which can cause things like Semantic coloring to "flash"
+    // when they update repeatedly as they load.
+    private static bool ProjectWorkspacePublishable(IProjectSnapshot? project)
+    {
+        return project?.ProjectWorkspaceState != null;
+    }
+
+    private void EnqueuePublish(IProjectSnapshot projectSnapshot)
+    {
+        _workQueue.AddWork((Project: projectSnapshot, Removal: false));
+    }
+
+    private void EnqueueRemoval(IProjectSnapshot projectSnapshot)
+    {
+        _workQueue.AddWork((Project: projectSnapshot, Removal: true));
+    }
+
+    private ValueTask ProcessBatchAsync(ImmutableArray<(IProjectSnapshot Project, bool Removal)> workItems, CancellationToken token)
+    {
+        foreach (var workItem in workItems.GetMostRecentUniqueItems(Comparer.Instance))
+        {
+            if (token.IsCancellationRequested)
+            {
+                return default;
+            }
+
+            if (workItem.Removal)
+            {
+                RemovePublishingData(workItem.Project);
+            }
+            else
+            {
+                ImmediatePublish(workItem.Project);
+            }
+        }
+
+        return default;
+    }
+
+    private void RemovePublishingData(IProjectSnapshot projectSnapshot)
+    {
+        if (_active)
```

Comment:
In theory it seems possible - we can add and remove a project before the server is started, right? :) Under debugger it seems to take a few seconds before server is started. Let me know if you disagree, otherwise I'll leave it in.

Reply:
> we can add and remove a project before the server is started, right?

Yes, but the event is only hooked up once the `StartSending` method is called, and `_active` is set before that, so whilst the project manager might see adds and removes, this service won't, right? Or am I reading something wrong?

Reply:
You are right, Remove shouldn't get called before we are active, my bad. Thanks!

Reply:
https://github.com/dotnet/razor/pull/10182/commits/339b3a39a76f906b493fc14806a3e494a01e8883